### ISSUE TITLE
Figure out what to do with harfbuzz dependency - bundle it?

### ISSUE THREAD

**Issue by tjyang** [tectonic-typesetting/tectonic#619]
- tectnoic info:
```
[x220@ipa01 tectonic]$ git remote -v
origin  https://github.com/tectonic-typesetting/tectonic.git (fetch)
origin  https://github.com/tectonic-typesetting/tectonic.git (push)
[x220@ipa01 tectonic]$ git log -1
commit b1973f17dbcb930efba57661a4b26edbc1117b7e (HEAD -> master, tag: continuous, origin/master, origin/HEAD)
Merge: afdeed4 53339f9
Author: Peter Williams <peter@newton.cx>
Date:   Mon Jul 6 11:32:56 2020 -0400

    Merge pull request #609 from tectonic-typesetting/dependabot/cargo/flate2-1.0.16

    build(deps): bump flate2 from 1.0.14 to 1.0.16
```
- warning messages from **cargo build**
```
[x220@ipa01 tectonic]$ cargo build
warning: tectonic/xetex-XeTeXFontInst.cpp: In function ‘hb_font_funcs_t* _get_font_funcs()’:
warning: tectonic/xetex-XeTeXFontInst.cpp:268:79: warning: ‘void hb_font_funcs_set_glyph_func(hb_font_funcs_t*, hb_font_get_glyph_func_t, void*, hb_destroy_func_t)’ is deprecated: Use 'hb_font_funcs_set_nominal_glyph_func and hb_font_funcs_set_variation_glyph_func' instead [-Wdeprecated-declarations]
warning:   268 |     hb_font_funcs_set_glyph_func                (funcs, _get_glyph, NULL, NULL);
warning:       |                                                                               ^
warning: In file included from /usr/include/harfbuzz/hb.h:34,
warning:                  from tectonic/xetex-core.h:28,
warning:                  from tectonic/xetex-XeTeXFontInst.cpp:42:
warning: /usr/include/harfbuzz/hb-deprecated.h:67:1: note: declared here
warning:    67 | hb_font_funcs_set_glyph_func (hb_font_funcs_t *ffuncs,
warning:       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: tectonic/xetex-XeTeXLayoutInterface.cpp: In function ‘unsigned int countFeatures(XeTeXFont, hb_tag_t, hb_tag_t)’:
warning: tectonic/xetex-XeTeXLayoutInterface.cpp:396:100: warning: ‘hb_bool_t hb_ot_layout_script_find_language(hb_face_t*, hb_tag_t, unsigned int, hb_tag_t, unsigned int*)’ is deprecated: Use 'hb_ot_layout_script_select_language' instead [-Wdeprecated-declarations]
warning:   396 |             if (hb_ot_layout_script_find_language(face, tableTag, scriptIndex, language, &langIndex) || language == 0) {
warning:       |                                                                                                    ^
warning: In file included from /usr/include/harfbuzz/hb-ot.h:34,
warning:                  from tectonic/xetex-core.h:29,
warning:                  from tectonic/xetex-XeTeXLayoutInterface.cpp:34:
warning: /usr/include/harfbuzz/hb-ot-deprecated.h:56:1: note: declared here
warning:    56 | hb_ot_layout_script_find_language (hb_face_t    *face,
warning:       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: tectonic/xetex-XeTeXLayoutInterface.cpp: In function ‘hb_tag_t getIndFeature(XeTeXFont, hb_tag_t, hb_tag_t, unsigned int)’:
warning: tectonic/xetex-XeTeXLayoutInterface.cpp:416:100: warning: ‘hb_bool_t hb_ot_layout_script_find_language(hb_face_t*, hb_tag_t, unsigned int, hb_tag_t, unsigned int*)’ is deprecated: Use 'hb_ot_layout_script_select_language' instead [-Wdeprecated-declarations]
warning:   416 |             if (hb_ot_layout_script_find_language(face, tableTag, scriptIndex, language, &langIndex) || language == 0) {
warning:       |                                                                                                    ^
warning: In file included from /usr/include/harfbuzz/hb-ot.h:34,
warning:                  from tectonic/xetex-core.h:29,
warning:                  from tectonic/xetex-XeTeXLayoutInterface.cpp:34:
warning: /usr/include/harfbuzz/hb-ot-deprecated.h:56:1: note: declared here
warning:    56 | hb_ot_layout_script_find_language (hb_face_t    *face,
warning:       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: tectonic/xetex-XeTeXLayoutInterface.cpp: In function ‘hb_unicode_funcs_t* _get_unicode_funcs()’:
warning: tectonic/xetex-XeTeXLayoutInterface.cpp:748:92: warning: ‘void hb_unicode_funcs_set_decompose_compatibility_func(hb_unicode_funcs_t*, hb_unicode_decompose_compatibility_func_t, void*, hb_destroy_func_t)’ is deprecated [-Wdeprecated-declarations]
warning:   748 |     hb_unicode_funcs_set_decompose_compatibility_func(ufuncs, _decompose_compat, NULL, NULL);
warning:       |                                                                                            ^
warning: In file included from /usr/include/harfbuzz/hb.h:34,
warning:                  from tectonic/xetex-core.h:28,
warning:                  from tectonic/xetex-XeTeXLayoutInterface.cpp:34:
warning: /usr/include/harfbuzz/hb-deprecated.h:158:1: note: declared here
warning:   158 | hb_unicode_funcs_set_decompose_compatibility_func (hb_unicode_funcs_t *ufuncs,
warning:       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
warning: tectonic/xetex-XeTeXLayoutInterface.cpp: In function ‘bool initGraphiteBreaking(XeTeXLayoutEngine, const uint16_t*, int)’:
warning: tectonic/xetex-XeTeXLayoutInterface.cpp:1012:78: warning: ‘gr_font* hb_graphite2_font_get_gr_font(hb_font_t*)’ is deprecated: Use 'hb_graphite2_face_get_gr_face' instead [-Wdeprecated-declarations]
warning:  1012 |     gr_font* grFont = hb_graphite2_font_get_gr_font(engine->font->getHbFont());
warning:       |                                                                              ^
warning: In file included from tectonic/xetex-XeTeXLayoutInterface.cpp:42:
warning: /usr/include/harfbuzz/hb-graphite2.h:45:1: note: declared here
warning:    45 | hb_graphite2_font_get_gr_font (hb_font_t *font);
warning:       | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    Finished dev [unoptimized + debuginfo] target(s) in 0.11s
[x220@ipa01 tectonic]$
```

**Comment by pkgw** [comment]
Thanks for reporting. These warnings remain because harfbuzz is an external dependency, and a lot of Tectonic platforms only provide older versions of the package, so we try to keep compatibility with them.

But, I have often thought that maybe we should use an internal copy of harfbuzz, or at least make it an option. I thought we must have had an open issue for that topic, but I'm not seeing one, so I'll declare that this issue will be it.

**Comment by pkgw** [comment]
Cf. #454 , which has been in draft state for a long time but would still be a good starting point if we make a change to require newer harfbuzz.

**Comment by pkgw** [comment]
That mention in #689 includes some helpful contextual info. I've been stymied in this idea for a long time because it seemed like it would be really hard to mix-n-match vendored and system library dependencies in Tectonic. But it turns out that there is [an established framework for "-sys" crates to expose C headers to one another](https://kornel.ski/rust-sys-crate#headers). (It's been around for a while, I just didn't understand it.) Combined with `cbindgen`, I think this gives us a path to being able to construct a build framework that provides a tractable option to vendor potential system deps such as Harfbuzz.

**Comment by pkgw** [comment]
Fixed in #724! We can now vendor harfbuzz.
